Added Team Loadouts to PlayerData

TessaHelenius opened the pull request:
Added:

TeamLoadOut (serializable)
- CustomCharacterListObject[] Slots (size 3)
- IsEmpty to help check if a slot has no characters

PlayerData:
- TeamLoadOut[] LoadOuts = { new(), new(), new() } = three saved teams -int SelectedLoadOut = 0 — current selection (0 = no loadout / “Current”) 
- ApplyLoadout(int index) = loads saved team into SelectedCharacterIds 
- SaveCurrentTeamToLoadout(int index) = writes SelectedCharacterIds to a slot with value copy (new instances) 
- OnCurrentTeamChanged_AutoSave() = when user edits team and a saved slot (1-3) is active, automatically save the position and persist it through Storefront 

(Decided to copy the values into a new CustomCharacterListObject when saving a loadout so we don’t share the same reference as SelectedCharacterIds)

- Added fourth loadout button on the left side of the existing ones

Commit:
Added Team Loadouts to PlayerData

Added:

TeamLoadOut (serializable)
-CustomCharacterListObject[] Slots (size 3)
-IsEmpty to help check if a slot has no characters

PlayerData:
-TeamLoadOut[] LoadOuts = { new(), new(), new() } = three saved teams
-int SelectedLoadOut = 0 — current selection (0 = no loadout / “Current”)
-ApplyLoadout(int index) = loads saved team into SelectedCharacterIds
-SaveCurrentTeamToLoadout(int index) = writes SelectedCharacterIds to a slot with value copy (new instances)
-OnCurrentTeamChanged_AutoSave() = when user edits team and a saved slot (1-3) is active, automatically save the position and persist it through Storefront
(Decided to copy the values into a new CustomCharacterListObject when saving a loadout so we don’t share the same reference as SelectedCharacterIds)

-Added fourth loadout button on the left side of the existing ones

Commit:
Fixed the commented parts of the code

-Moved TeamLoadOut-class to its own script
-Changed hardcoded values to use LoadOuts.Length
-Added Debug.LogError messages on invalid loadout index in ApplyLoadout and SaveCurrentTeamToLoadout
-Removed the ServerID check

Commit:
Add loadout handling and PlayerData integration between ModelController and LoadOutController

-Awake() Gets the ModelController component and subscribes to the OnPlayerDataReady event to receive playerdata once its loaded

-Start() Binds each loadout button in the list to the OnPressLoadout method. When button is clicked, it passes its corresponding index to handle loadout selection or saving

-HaldePlayerDataReady() This is called when playerdata becomes available, it stores the reference to the loaded player and updates the button states

-OnPressLoadout() Handles what happens when a loadout button is pressed, also updates button states and saves the playerdata afterward

-RefreshButtons() Updates the button interactability based on which loadout is currently active, and the active loadout button is disabled to indicate selection

In ModelController.cs i added:

-Event to notify LoadOutController when playerdata has been successfully loaded

-Public readonly property to expose the current PlayerData instance

-OnPlayerDataReady?.Invoke(_playerData) that invokes the OnPlayerDataReady event once the player data has finished loading

Commit:
Loadout: wire ModelController via SerializeField, pass PlayerData into OnPressLoadout, add Storefront fallback fetch

- Exposed ModelController to the scene via SerializeField so it can be assigned from the editor

- Changed OnPressLoadout to accept PlayerData as a parameter and added a null check

- If PlayerData is not available, added a fallback fetch from Storefront in LoadOutController to ensure the latest player info is retrieved before applying a loadout.

Commit:
Merge branch 'main' into UI/DefenceGallery/Feature/TeamLoadOuts

Commit:
Added loadout save prompt, “slot 0” popup, and cleaned code a bit

- Moved new PlayerData.cs methods to the end of the file for clarity

- Added a confirmation popup that asks whether to save the current loadout to the current slot before switching

- Hooked up a new popup window that opens from the “0” button

- Implemented a LoadoutPopup UI with three character slots and a save button (just UI, not working yet)

Commit:
Merge branch 'main' into UI/DefenceGallery/Feature/TeamLoadOuts

Commit:
Added SelectedCharacters slots to LoadoutPopup, adjusted its UI, and disabled autosave

- Implemented new SelectedCharacters slots in the LoadoutPopup window

- Updated and refined the LoadoutPopup UI layout for better usability

- Disabled the autosave functionality as it's not needed right now

Commit:
Add loadout save confirmation, selectable row backgrounds, and UI improvements etc.

- Added a confirmation popup (LoadoutSaveConfirmPopup) that appears before saving a loadout

- Hooked up all “save” buttons to open the confirmation popup instead of saving immediately

- Added logic in LoadoutRowController to track the active loadout row and handle confirmed saves

- Added clickable dark-blue row backgrounds to allow selecting a saved loadout directly

- Hooked up background buttons to call OnRowBackgroundClicked and update the active loadout selection

- Added LoadoutEditorOpener helper for opening the edit panel through SignalBus

- Improved (or at least tried) slot button initialization and error handling in LoadoutRowController

Commit:
Merge branch 'main' into UI/DefenceGallery/Feature/TeamLoadOuts

BillTheBeast committed:
SelectedCharacter slot details get properly enabled again.

Commit:
Added inline/popup loadout mode switch and button variants

- Added new LoadoutButtons prefab and a LoadoutButtons Variant prefab for the two loadout layouts

- Added “Popup Loadout” and “Inline Loadout” buttons to the tab ribbon to switch between layouts

- Implemented a small LoadoutModeSwitcher script that toggles visibility between the default loadout buttons and the variant based on the selected mode

Commit:
Merge branch 'main' into UI/DefenceGallery/Feature/TeamLoadOuts

BillTheBeast committed:
Merge branch 'main' into UI/DefenceGallery/Feature/TeamLoadOuts

Commit:
Added Loadout 2 dynamic loadout system with runtime slot creation and automatic button resizing

-Implemented the new Loadout 2 variant where players can add additional loadout slots by pressing the plus button in the UI

-Added method AddNewLoadoutSlot() to LoadOutController, which creates a new loadout button, inserts it before the plus button, and wires up the correct click logic

-ScrollView currently displays 4 loadouts plus the add (“+”) button

Added new script InlineLoadoutButtonSizer.cs that:

-Automatically calculates the correct button size based on the viewport width, layout padding, and spacing

-Ensures that a fixed number of loadout buttons fit visibly in the horizontal ScrollView

-Resizes both width and height of each button to maintain a consistent square layout

Also did some general UI adjustments and button refresh logic improvements to support dynamically created loadouts and added mode switch highlighting for Popup/Inline loadout buttons, making the selected mode button appear dark (disabled state), matching the style of the loadout buttons

Commit:
Added popup loadout sync, hid inline add-button when max slots added, and fixed inline button issue

- Added functionality so the first three loadouts shown on the popup loadout main page are now also displayed correctly in popup rows 1–3

- Added automatic hiding of the "+" button in the inline loadout view when the maximum number of loadouts has been created

- Fixed an issue where inline loadout buttons had oversized raycast areas, causing clicks to select the button to the right instead of the intended one

Commit:
Unified loadout lists, refined inline/popup behavior, and updated popup interaction

- Hidden the inline loadout “+” button once the 8th loadout becomes visible

- Extracted copy button into its own prefab

- Inline and popup loadouts now share the same unified 8-slot list

- Added SwipeBlocker to the popup loadout window

- Updated popup loadout: rename “Save” button to “Change” and connect it to open the character selection view

Commit:
Added loadout-aware character editing and automatic loadout saving in EditingPopup & ModelController

- Added support for opening the character editing popup from a specific loadout

- Modified EditingPopup to track whether it was opened via loadout editing

- Added new signal handling so character changes update the correct loadout instead of only SelectedCharacterIds

- Implemented HandleLoadoutCharacterChanged in ModelController to write changes directly into the chosen loadout

- Synced logic so that when the edited loadout is the currently selected one, SelectedCharacterIds and server data are also updated

- Ensured automatic saving occurs for the correct loadout when editing is closed

BillTheBeast committed:
Fixed loadout popup incorrectly updating previously modified loadout visually

Commit:
Dynamic inline loadout buttons (add/hide/reorder) + prevent scroll jump

- Refactored inline loadout button handling in LoadOutController to support dynamic slots cleanly

- Added a central RegisterButton() helper to:
  - assign click handlers consistently (OnPressLoadout with correct loadoutIndex)
  - maintain _buttonMap (Button -> LoadoutIndex mapping)
  - update TMP label text to match the loadout number

- Implemented CreateDynamicButton() to instantiate new loadout buttons from a template, register them, track them in _dynamicButtons, and optionally trigger reorder.
- Implemented FindNextHiddenEmptyLoadoutIndex() + UpdateAddButtonVisibility() so:
  - the “+” button is shown only when there is at least one hidden empty loadout (starting from index 4)
  - pressing “+” adds the next available empty slot

- Implemented ReorderInlineButtonsByLoadoutIndex() to keep inline loadout buttons ordered (4,5,6,...) and always place “+” after the last loadout button
- Updated BuildInlineButtons() rebuild flow to stop the visible “row jump” by caching and restoring ScrollRect/content position and forcing layout rebuild before restoring

- Removed ActiveRowIndex from LoadoutRowController. The selected row can be derived directly from PlayerData.SelectedLoadOut so a separate index was unnecessary

Commit:
Fix(loadouts): prevent accidental save when selecting empty loadout in inline mode

- Fixed an issue where selecting an empty loadout incorrectly saved the current team into that loadout

Commit:
Fix loadout selection, inline cleanup timing, and scroll jump when adding loadouts

- Fixed an issue where the previously active popup loadout could visually override
  another loadout after reopening the view

- When adding a new loadout via the "+" button in inline mode, the newly created loadout
  is now automatically selected immediately

- Changed the timing of inline loadout cleanup logic so that empty loadout buttons
  (loadouts 4+) are only removed when the view is exited (OnDisable) and rebuilt
  when the view is re-entered (OnEnable)

- Added logic to ensure that if all loadout slots are empty but characters are currently
  selected (SelectedCharacters), the selected characters are automatically stored into the first loadout slot

- Fixed the horizontal scroll "jump" when adding new inline loadouts via the "+"
  button

Commit:
Fix event unsubscription in EditingPopup OnDestroy

- Fixed a bug where EditingPopup was subscribing to OnDefenceGalleryEditPanelRequestedForLoadout in OnDestroy instead of unsubscribing

BillTheBeast commented on Assets/MenuUi/Scripts/DefenceScreen/CharacterGallery/LoadoutRowController.cs:
```diff
@@ -0,0 +1,293 @@
+    /// <summary>
+    /// Saves current team into this row's saved slot and exits "live" preview
+    /// </summary>
+    public void SaveCurrentTeamIntoThisLoadout()
```
Onko tarpeellinen enää?

Commit:
Fix loadout change not being saved to server & refactor and clean the code

- Fixed an issue where loadout changes were not persisted to the server

- Removed or commented out unused code sections

- Refactored code

BillTheBeast committed:
Made the BattleCharacter saving code into its own method.

Switched ModelController and LoadoutControllers to use it instead.
Loadout now saves properly to server when using the loadout popup.

BillTheBeast committed:
Merge branch 'main' into UI/DefenceGallery/Feature/TeamLoadOuts
